graphql_compiler/cost_estimation/statistics.py

```diff
@@ -0,0 +1,207 @@
+        return self.__str__()
+
+    @abstractmethod
+    def get_class_count(self, class_name):
+        """Return how many objects have, or inherit, the given class name.
+
+        Args:
+            class_name: str, either a vertex class name or an edge class name defined in the
+                        GraphQL schema.
+
+        Returns:
+            - Number of objects with, or inheriting the given class name
+
+        Raises:
+            AssertionError, if statistic for class_name was not provided.
+        """
+        raise NotImplementedError()
+
+    @abstractmethod
+    def get_edge_count_between_vertex_pair(
+        self, vertex_out_class_name, vertex_in_class_name, edge_class_name
+    ):
+        """Return the edge count between two vertices of a given type using an edge of a given type.
+
+        Edge traversal cost between two vertices using an edge that connects the vertices' using
+        their abstract inherited types is difficult to estimate, so we provide an interface for
+        this statistic.
+
+        Args:
+            vertex_out_class_name: str, vertex class name.
+            vertex_in_class_name: str, vertex class name.
+            edge_class_name: str, edge class name.
+
+        Returns:
+            - The count of edges if the statistic exists
+            - None otherwise
+        """
+        return None
+
+    @abstractmethod
+    def get_domain_count(self, vertex_name, field_name):
+        """Return the number of distinct values the vertex field has over all vertex instances.
+
+        This statistics helps estimate the result size of @filter directives used to restrict
+        values using equality operators like '=', '!=', and 'in_collection'.
+
+        Preconditions:
+        1. field_name must be a valid vertex field of vertex_name
+
+        Args:
+            vertex_name: str, name of a vertex.
+            field_name: str, name of a vertex field.
+
+        Returns:
+            - Distinct value count if the statistic exists
+            - None otherwise
+        """
+        return None
+
+    @abstractmethod
+    def get_histogram(self, vertex_name, field_name):
+        """Return a histogram for the given vertex field, providing statistics about range values.
+
+        This statistics helps estimate the result size of @filter directives used to restrict
+        numbers using inequality operators like '>', '<', and 'between'.
+
+        Preconditions:
+        1. field_name must be a valid vertex field of vertex_name
+
+        Args:
+            vertex_name: str, name of a vertex.
+            field_name: str, name of a vertex field.
+
+        Returns:
+            - A histogram as a list of tuples with 3 elements each if the statistic exists.
+              Each tuple represents a bin as (start, end, elementCount), where start and end are
+              the interval boundaries (inclusively), and elementCount is the bin element count.
```

Comment:
Including the elements at both boundaries in each bin might lead to overcounting. Why not have the count be inclusive of the elements at the start boundary and not the end boundary?